### PR TITLE
Current active development branch for low level CAN

### PR DESCRIPTION
**Pull request by vkottler**
Creating a PR so we can hold a discussion here instead of on Slack about the code

### CODE REVIEW

File: embedded/drivers/st/stm32l4xx_hal_can.c [hunk 1]
```diff
@@ -1942,6 +1942,29 @@ HAL_StatusTypeDef HAL_CAN_ResetError(CAN_HandleTypeDef *hcan)
...
+void CAN1_TX_IRQHandler(void){
+	HAL_CAN_IRQHandler();
+}
+
+void CAN1_TX_IRQHandler(void){
+	HAL_CAN_IRQHandler();
+}
+
+/* BADGERLOOP ADDED END 05/24/2018 */
+
```
Comment (vkottler) [inline embedded/drivers/st/stm32l4xx_hal_can.c]: Does this compile? This is what I see as the prototype for `HAL_CAN_IRQHandler`:

```
void HAL_CAN_IRQHandler(CAN_HandleTypeDef *hcan);
```

Which would hold all of the config information so this function can do the right thing

File: embedded/common/board.c [hunk 2]
```diff
@@ -50,6 +53,34 @@ int periph_init(void) {
...
+        HAL_NVIC_SetPriority(CAN1_RX0_IRQn,0,0);
+	HAL_NVIC_SetPriority(CAN1_RX1_IRQn,0,0);
+	HAL_NVIC_SetPriority(CAN1_SCE_IRQn,0,0);
+
+        HAL_NVIC_EnableIRQ(CAN1_TX_IRQn);
+        HAL_NVIC_EnableIRQ(CAN1_RX0_IRQn);
+        HAL_NVIC_EnableIRQ(CAN1_RX1_IRQn);
+        HAL_NVIC_EnableIRQ(CAN1_SCE_IRQn);
+
+
```
Comment (vkottler) [inline embedded/common/board.c]: I would not enable ISRs unless it was preceded by copying working, blocking code into an ISR or helper / aggregate ISR

**Comment by vkottler** [comment]
Definitely don't worry about making similar changes to all of the files in `app`.

You are doing active development so just put them in `dev.c` and build and flash `dev`.

I see that [`0f53ec3`](https://github.com/badgerloop-software/pod/pull/13/commits/0f53ec3ada18e8f2562b9a47b2ecfea5845dad94) has a bunch of identical changes which makes it hard to review

**Comment by vkottler** [comment]
@rcastle1 @ethanlink17 any comments on what functionality is expected or what might be missing?

I can try it out on my breadboard setup tomorrow

**Comment by rcastle1** [comment]
Ok, will just do dev work in  `dev.c`, however, we needed to update all of the files in `app` as the definition of periph_init() changed to peiph_init(CAN_HandleTypeDef hcan) and required the CAN Globals to compile.

Goals:

- ~~First stepping through the CAN_INIT() to see what is causing the infinite loop, likely an error handler~~
- Once we are initializing ok, for local dev we can shoot to read CAN messages from the BMS. That requires the baud rate to be set at 250kbps.
- Remotely, if two boards are set up with the CAN Transceivers I think the goal would be to get two talking. I looked at the [MCP2561](http://ww1.microchip.com/downloads/en/DeviceDoc/20005167C.pdf) datasheet and didn't see anything about a single node loopback mode. 
- Write the higher level can_read() can_write() can_query() functions and update the serial interface commands to be able to easily interface with the BMS and RMS 
- Once we know we can send/read messages at 250kbps, the motor controller set up and we should be able to command the motor.

**Comment by rcastle1** [comment]
CAN_INIT() issue was likely just a flashing issue. Boards can flash an initialize CAN with code at [804099e](https://github.com/badgerloop-software/pod/pull/13/commits/804099e1df4bbff9649e4ea8a93fe67cb839b47b) after reflashing.

**Comment by vkottler** [comment]
Loopback mode has nothing to do with transceivers. Instead of sending anything over the wire it literally muxes its own output to its own input in hardware (microcontroller's CAN interface).

I wouldn't use global variables for the CAN configuration. It just unnecessarily eats up RAM and only gets used once.

If you're trying to set hard-coded parameters that differ per module, use `#define`.

Nothing redundant/equivalent should live in two different locations

**Comment by rcastle1** [comment]
ideas on how to check if baud rate is correct?

**Comment by vkottler** [comment]
I would set the logic analyzer to the baud you expect and see if it can decode it. It puts dots where it tries to reconcile the value on the wire

Taking an oscilloscope measurement works decently as well

**Comment by vkottler** [comment]
Open a PR when this code is ready and tested.

**Pull request closed without merging**